compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -359,32 +372,40 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         dest: &PlaceTy<'tcx, M::PointerTag>,
     ) -> InterpResult<'tcx> {
         trace!("Unsizing {:?} of type {} into {:?}", *src, src.layout.ty, cast_ty.ty);
-        match (&src.layout.ty.kind(), &cast_ty.ty.kind()) {
-            (&ty::Ref(_, s, _), &ty::Ref(_, c, _) | &ty::RawPtr(TypeAndMut { ty: c, .. }))
-            | (&ty::RawPtr(TypeAndMut { ty: s, .. }), &ty::RawPtr(TypeAndMut { ty: c, .. })) => {
+        let just_metadata = self.tcx.lang_items().just_metadata();
+        match (src.layout.ty.kind(), cast_ty.ty.kind()) {
+            (ty::Ref(_, s, _), ty::Ref(_, c, _) | ty::RawPtr(TypeAndMut { ty: c, .. }))
+            | (ty::RawPtr(TypeAndMut { ty: s, .. }), ty::RawPtr(TypeAndMut { ty: c, .. })) => {
                 self.unsize_into_ptr(src, dest, *s, *c)
             }
-            (&ty::Adt(def_a, _), &ty::Adt(def_b, _)) => {
+            (ty::Adt(def_a, _), ty::Adt(def_b, _)) if def_a.is_box() || def_b.is_box() => {
                 assert_eq!(def_a, def_b);
-                if def_a.is_box() || def_b.is_box() {
-                    if !def_a.is_box() || !def_b.is_box() {
-                        span_bug!(
-                            self.cur_span(),
-                            "invalid unsizing between {:?} -> {:?}",
-                            src.layout.ty,
-                            cast_ty.ty
-                        );
-                    }
-                    return self.unsize_into_ptr(
-                        src,
-                        dest,
-                        src.layout.ty.boxed_ty(),
-                        cast_ty.ty.boxed_ty(),
+                if !def_a.is_box() || !def_b.is_box() {
+                    span_bug!(
+                        self.cur_span(),
+                        "invalid unsizing between {:?} -> {:?}",
+                        src.layout.ty,
+                        cast_ty.ty
                     );
                 }
-
+                self.unsize_into_ptr(src, dest, src.layout.ty.boxed_ty(), cast_ty.ty.boxed_ty())
+            }
+            (ty::Adt(def_a, substs_a), ty::Adt(def_b, substs_b))
```

Comment:
If we want to pursue unsizing working directly with `<T as Pointee>::Metadata`, it would replace this arm checking for `JustMetadata<T>` with checking for the projection. The code *should* be nearly the same, since `JustMetadata<T>` is just lifting the projection to an `Adt` ty kind.

Reply:
Projections tend to get normalized away though,  so that does not sound like a great idea.